go/arrow/bitutil/bitutil_go121.go

```diff
@@ -0,0 +1,218 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.21
+// +build go1.21
+
+package bitutil
+
+import (
+	"math"
+	"math/bits"
+	"unsafe"
+
+	"github.com/apache/arrow/go/v15/arrow/memory"
+)
+
+var (
+	BitMask        = [8]byte{1, 2, 4, 8, 16, 32, 64, 128}
+	FlippedBitMask = [8]byte{254, 253, 251, 247, 239, 223, 191, 127}
+)
+
+// IsMultipleOf8 returns whether v is a multiple of 8.
+func IsMultipleOf8(v int64) bool { return v&7 == 0 }
+
+// IsMultipleOf64 returns whether v is a multiple of 64
+func IsMultipleOf64(v int64) bool { return v&63 == 0 }
+
+func BytesForBits(bits int64) int64 { return (bits + 7) >> 3 }
+
+// NextPowerOf2 rounds x to the next power of two.
+func NextPowerOf2(x int) int { return 1 << uint(bits.Len(uint(x))) }
+
+// CeilByte rounds size to the next multiple of 8.
+func CeilByte(size int) int { return (size + 7) &^ 7 }
+
+// CeilByte64 rounds size to the next multiple of 8.
+func CeilByte64(size int64) int64 { return (size + 7) &^ 7 }
```

Comment:
Rather than duplicating all of these, can you leave `bitutil.go` with all of the common stuff and make new files that use the build constraints which *only* consists of the stuff which is different for each? I'd prefer to minimize and reduce the duplication here